### PR TITLE
add source_as function

### PR DESCRIPTION
**Pull request by daviddalpiaz**
Added a `source_as` function.

Thanks to @barbehenna for the clever `lsf.str` trick. 

With this added function, test scripts will be both safer and clearer.

As an example, a question that asks students to write a function will now look like:

```r
# load packages and extensions
library(tinytest)
using(ttdo)

# source student and reference files
std = plr::source_as(file = "/grade/student/submission.R")
ref = plr::source_as(file = "/grade/tests/solution.R")

## @title Does the function return the correct type with default inputs?
## @score 1
expect_equal_with_diff(typeof(std$fun()), typeof(ref$fun()))

## @title Does the function return the correct object when x = 1?
## @score 1
expect_equal_with_diff(std$fun(x = 1), ref$fun(x = 1))
```

When students are asked to create objects:

```r
# load packages and extensions
library(tinytest)
using(ttdo)

# source student and reference files
std = plr::source_as(file = "/grade/student/submission.R")
ref = plr::source_as(file = "/grade/tests/solution.R")

## @title Does the vector a have the correct length?
## @score 1
expect_equal_with_diff(length(std$a), length(ref$a))

## @title Does the vector b have the correct length?
## @score 1
expect_equal_with_diff(length(std$b), length(ref$b))

## @title Does the vector c have the correct length?
## @score 1
expect_equal_with_diff(length(std$c), length(ref$c))
```

Open to suggestions about the name of the function. 

Currently leaving the existing functions `source_and_eval_safe` and `source_and_eval_safe_with_hiding` as-is so as not to break any existing questions. But long term, we should consider using the `source_as` function within them to bring added security. However, assuming we add `source_as`, future questions should leverage `source_as`.

### CODE REVIEW

**Comment by daviddalpiaz** [comment]
Tagging @eddelbuettel in case he isn't subscribed to changes here.

**Comment by eddelbuettel** [comment]
One thought I just had:  In

```r
# source student and reference files
std = plr::source_as(file = "/grade/student/submission.R")
ref = plr::source_as(file = "/grade/tests/solution.R")
```

the paths are a given. May wrap via something alike (riffing, untested)

```r
source_as_base <- function(filename, basepath) {  source_as(file.path(basepath, filename) }
source_as_student <- function(name) source_as_base(name, "/grade/student")
source_as_reference <- function(name) source_as_base(name, "/grade/tests")

# ...

std <- plr::source_as_student("submission.R")
ref <- plr::source_as_reference("solution.R")   # or "answer.R" or ...
```

**Review by eddelbuettel — APPROVED** [review]
Looks good to me as discussed on Slack.  Let's merge and see how it fares!

One minor suggestion / idea in comment.

**Comment by daviddalpiaz** [comment]
Oh! Love that idea. How would you feel about?

```r
source_as_base <- function(filename, basepath) {  source_as(file.path(basepath, filename) }
source_as_std <- function(name) source_as_base(name, "/grade/student")
source_as_ref <- function(name) source_as_base(name, "/grade/tests")
```

Silly little difference, but you can sort of see why here:

```r
std <- plr::source_as_std("submission.R")
ref <- plr::source_as_ref("solution.R")   # or "answer.R" or "ans.R" :)
```

**Comment by eddelbuettel** [comment]
Yep, looks good.  Maybe rename the actual work `source_as_name_and_dir` or something, or even make it a hidden function.

**Comment by daviddalpiaz** [comment]
Just took a stab at something like that. Now we have:

- `source_as_uid(file, uid = "ag")`: Generic function that can source any arbitrary file as any user.
- `source_std(name)`: Source a file from the `/grade/student/` directory, only as user `"ag"`.
- `source_ref(name, uid = "ag")`: Source a file from the `/grade/tests/` directory, as any user, but defaulted to `"ag"`.

All three are exported. Users would be encouraged to use the `_std` and `_ref` variants, but I can imagine use cases where users should have access to `source_as_uid`.

I'm unsure what privileges should be granted within `source_ref`, but I imagine there are cases where it might need to be higher than `"ag"`.

File: R/source.R [hunk 1]
```diff
@@ -50,3 +59,15 @@ source_as <- function(file, uid="ag") {
...
+source_std <- function(name) {
+  full_path = file.path("/grade/student", name)
+  source_as_uid(file = full_path)
+}
+
+#' @rdname source_as_uid
+source_ref <- function(name) {
+  full_path = file.path("/grade/tests", name)
+  source_as_uid(file = full_path)
+}
```
Comment (eddelbuettel) [inline R/source.R:73]: That doesn't have the same elegance I had in mind, and it cloaks the uid argument.  Not worth fighting to death so just leave it ...

Using `=` for assigment is still icky. :stuck_out_tongue_winking_eye:

Comment (daviddalpiaz) [reply R/source.R:73]: Oh gosh. No comment on the which operator is better, but I die inside when it isn't consistent. I'll clean that up. Habits and such...

Would your preference be that the `_std` variant mirrors the `_ref` variant and takes a `uid` argument? I suppose keeping them parallel makes sense. I wasn't sure if part of your suggestion was to hide the `uid` argument in these helper variants based on your code above.

Comment (eddelbuettel) [reply R/source.R:73]: In the general case I have often found regretting later to not allow (all) args through.  Now, here and with PL, we could argue `ag` is a constant so it is probably fine.

**Review by eddelbuettel — APPROVED** [review]
Love it

**Comment by daviddalpiaz** [comment]
@barbehenna Can you give this a quick glance when you get a chance before we merge?

**Comment by barbehenna** [comment]
This looks great!

**Review by barbehenna — CHANGES_REQUESTED** [review]
typo?

File: R/source.R [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    temp_env <- new.env()
+    source(file, temp_env)
+    temp_env
+  }, uid = uid)
+
+  # restore original file permissions
+  Sys.chmod(file, mode=oldmode)
+
+  for (f in lsf.str(envir = sourced_env)) {
+    body(sourced_env[[f]]) <- parse(text = paste("eval_safe(", as.expression(body(env[[f]])), ", uid =", uid, ")"))
```
[the remark this answers is not in the document]
Comment (barbehenna) [reply R/source.R:55]: It also would be nice to add `importFrom("utils", "lsf.str")` to the `NAMESPACE`

[the remark this answers is not in the document]
Comment (daviddalpiaz) [reply R/source.R]: Fixed these as well as a small documentation issue.

**Review by eddelbuettel — APPROVED** [review]

**Comment by daviddalpiaz** [comment]
@barbehenna @eddelbuettel Any objections to merging this now?

**Comment by barbehenna** [comment]
Looks good to me!

**Comment by eddelbuettel** [comment]
(Had dinner guests from out of town whom I just dropped off near U of C but looks good so ... merged!)

**Pull request merged**